### PR TITLE
Adding two new os.path filters and its docs

### LINKED ISSUE #78675: More os.path filters
### Summary

Please add more filters based on os.path, specifically: `os.path.commonpath`, `os.path.normpath`. It is cheap, but would be handy in such tasks as validation, archive management, programmatic path generation. Can replace a lot of loops and regex filters.

### Issue Type

Feature Idea

### Component Name

core

### Additional Information

none

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by shivamdurgbuns**
Signed-off-by: Shivam Durgbuns <sdurgbun@redhat.com>

##### SUMMARY

Adding two new filters based on os.path, specifically: `os.path.commonpath`, `os.path.normpath`
Fixes #78675 

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

core

### CODE REVIEW

File: docs/docsite/rst/playbook_guide/playbooks_filters.rst [hunk 1]
```diff
@@ -2021,6 +2021,18 @@ To join one or more path components:
 
 .. versionadded:: 2.10
 
+.. code-block:: yaml+jinja
+
+    {{ path | commonpath }}
+
+To get Normalize path, eliminating double slashes, etc. (new in version 2.10):
```
Comment (webknjaz) [inline docs/docsite/rst/playbook_guide/playbooks_filters.rst]: 1) The next release is 2.14 and that's when it might be added
2) `.. versionadded::` Sphinx directive must be used for such information, consistent with other places.
3) Trailing colon is unnecessary.

File: lib/ansible/plugins/filter/core.py [hunk 2]
```diff
@@ -565,6 +565,16 @@ def path_join(paths):
         raise AnsibleFilterTypeError("|path_join expects string or sequence, got %s instead." % type(paths))
 
 
+def commonpath(paths):
+    '''
+    takes a list of paths and it returns the longest common path from the given list.
+    '''
+    if is_sequence(paths):
```
Comment (webknjaz) [inline lib/ansible/plugins/filter/core.py]: It's best to reverse the check, turning it into a guard expression.
Doing
```diff
-    if is_sequence(paths):
+    if not is_sequence(paths):
+        raise ...
```
would allow to avoid indenting the positive case unnecessarily (as in, the else-block would be removed).

File: lib/ansible/plugins/filter/core.py [hunk 3]
```diff
@@ -565,6 +565,16 @@ def path_join(paths):
         raise AnsibleFilterTypeError("|path_join expects string or sequence, got %s instead." % type(paths))
 
 
+def commonpath(paths):
+    '''
+    takes a list of paths and it returns the longest common path from the given list.
+    '''
```
Comment (webknjaz) [inline lib/ansible/plugins/filter/core.py]: I understand that this has been copied from the code above, but those aren't the best examples of writing docstrings, unfortunately.

Per [PEP 257](https://peps.python.org/pep-0257/), the first line should be an imperative-style active-voice sentence. Starting with a capital letter, ending with a trailing period, fitting in 72 chars. If it doesn't, maybe it's too wordy and some of it needs to elsewhere.
If there's important details to add, those should go into an extended description section in a separate paragraph, separated with one empty line after the title.

Here's one way of solving this:
```suggestion
    """Retrieve the longest common path from the given list.

    :param paths: A list of file system paths.
    :type paths: List[str]

    :returns: The longest common path.
    :rtype: str
    """
```

(the param definitions here follow Sphinx's markup declared @ https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists)

File: lib/ansible/plugins/filter/core.py [hunk 4]
```diff
@@ -600,6 +610,8 @@ def filters(self):
             'win_basename': partial(unicode_wrap, ntpath.basename),
             'win_dirname': partial(unicode_wrap, ntpath.dirname),
             'win_splitdrive': partial(unicode_wrap, ntpath.splitdrive),
+            'commonpath': commonpath,
+            'normpath': partial(unicode_wrap, os.path.normpath),
```
Comment (webknjaz) [inline lib/ansible/plugins/filter/core.py]: Why is only one of these wrapped?

**Comment by webknjaz** [comment]
@shivamdurgbuns you may want to split this into two PRs. Each would need a changelog fragment.

File: docs/docsite/rst/playbook_guide/playbooks_filters.rst (hunk 1 above)
Comment (bcoca) [reply docs/docsite/rst/playbook_guide/playbooks_filters.rst]: don't bother with this page, i'm removing most content soon, see this as the 'new way' to document filters/tests https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/filter/dirname.yml

Comment (shivamdurgbuns) [reply docs/docsite/rst/playbook_guide/playbooks_filters.rst]: So do I need to make the changes or not?

File: lib/ansible/plugins/filter/core.py (hunk 2 above)
Comment (shivamdurgbuns) [reply lib/ansible/plugins/filter/core.py]: Understood, will make this change.

File: lib/ansible/plugins/filter/core.py (hunk 3 above)
Comment (shivamdurgbuns) [reply lib/ansible/plugins/filter/core.py]: No problem, I thought we aren't using descriptive docstrings for the functions here, thus copied the same as above to keep the singularity in code. Will make this change.

File: lib/ansible/plugins/filter/core.py (hunk 4 above)
Comment (shivamdurgbuns) [reply lib/ansible/plugins/filter/core.py]: wrapped as in in the partial function? actually, for the `commonpath` we will need to provide a list of paths, and I can see that the `unicode_wrap` only returns a string as an output, thus I didn't wrap it.
@webknjaz Is there any better way to do that?

**Comment by shivamdurgbuns** [comment]
> @shivamdurgbuns you may want to split this into two PRs. Each would need a changelog fragment.

Have created a new PR here: https://github.com/ansible/ansible/pull/78798

File: docs/docsite/rst/playbook_guide/playbooks_filters.rst (hunk 1 above)
Comment (bcoca) [reply docs/docsite/rst/playbook_guide/playbooks_filters.rst]: not the ones on playbooks_filters page, but you need to create the .yml documentation

Comment (shivamdurgbuns) [reply docs/docsite/rst/playbook_guide/playbooks_filters.rst]: Understood will make the change

Comment (shivamdurgbuns) [reply docs/docsite/rst/playbook_guide/playbooks_filters.rst]: @bcoca I have added the .yml documentation for the commonpath, could you please review it, it's my first.

File: lib/ansible/plugins/filter/commonpath.yml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+DOCUMENTATION:
+  name: commonpath
+  author: Shivam Durgbuns
+  version_added: "historical"
```
Comment (bcoca) [inline lib/ansible/plugins/filter/commonpath.yml]: 2.15 at this point

Comment (shivamdurgbuns) [reply lib/ansible/plugins/filter/commonpath.yml]: Have changed it.
